lib/Service/ProvisioningService.php

```diff
@@ -0,0 +1,174 @@
+		$displaynameAttribute = $this->providerService->getSetting($providerId, ProviderService::SETTING_MAPPING_DISPLAYNAME, 'name');
+		$userName = $idTokenPayload->{$displaynameAttribute} ?? null;
+		$quotaAttribute = $this->providerService->getSetting($providerId, ProviderService::SETTING_MAPPING_QUOTA, 'quota');
+		$quota = $idTokenPayload->{$quotaAttribute} ?? null;
+
+		$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_UID, $idTokenPayload, $sub);
+		$this->eventDispatcher->dispatchTyped($event);
+
+		$backendUser = $this->userMapper->getOrCreate($providerId, $event->getValue());
+		$this->logger->debug('User obtained from the OIDC user backend: ' . $backendUser->getUserId());
+
+		$user = $this->userManager->get($backendUser->getUserId());
+		if ($user === null) {
+			return null;
+		}
+
+		// Update displayname
+		if (isset($userName)) {
+			$newDisplayName = mb_substr($userName, 0, 255);
+			$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_DISPLAYNAME, $idTokenPayload, $newDisplayName);
+		} else {
+			$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_DISPLAYNAME, $idTokenPayload);
+		}
+		$this->eventDispatcher->dispatchTyped($event);
+		$this->logger->debug('Displayname mapping event dispatched');
+		if ($event->hasValue()) {
+			$newDisplayName = $event->getValue();
+			if ($newDisplayName != $backendUser->getDisplayName()) {
+				$backendUser->setDisplayName($newDisplayName);
+				$backendUser = $this->userMapper->update($backendUser);
+			}
+		}
+
+		// Update e-mail
+		$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_EMAIL, $idTokenPayload, $email);
+		$this->eventDispatcher->dispatchTyped($event);
+		$this->logger->debug('Email mapping event dispatched');
+		if ($event->hasValue()) {
+			$user->setEMailAddress($event->getValue());
+		}
+
+		$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_QUOTA, $idTokenPayload, $quota);
+		$this->eventDispatcher->dispatchTyped($event);
+		$this->logger->debug('Quota mapping event dispatched');
+		if ($event->hasValue()) {
+			$user->setQuota($event->getValue());
+		}
+
+		// Update groups
+		if ($this->providerService->getSetting($providerId, ProviderService::SETTING_GROUP_PROVISIONING, '0') === '1') {
+			$this->provisionUserGroups($user, $providerId, $idTokenPayload);
+		}
+
+		return $user;
+	}
+
+	public function provisionUserGroups(IUser $user, int $providerId, object $idTokenPayload): void {
+		$groupsAttribute = $this->providerService->getSetting($providerId, ProviderService::SETTING_MAPPING_GROUPS, 'groups');
+		$groupsData = $idTokenPayload->{$groupsAttribute} ?? null;
+
+		$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_GROUPS, $idTokenPayload, json_encode($groupsData));
+		$this->eventDispatcher->dispatchTyped($event);
+		$this->logger->debug('Group mapping event dispatched');
+
+		if ($event->hasValue()) {
+			$groups = json_decode($event->getValue());
```

Comment:
the groupsData, like any other extracted field from the token, is dispatched to an event that may modify and map the data. the event only allows setting and returning a value as string so it has to be encoded somehow. we chose json since openid tokens uses jwt and it is easy to convert.

Reply:
:+1: